File: fixtures/scale/dart/medium/lib/funcs/stats.dart
import 'dart:math' as math;

import '../eval/function_registry.dart';

void registerStats(FunctionRegistry registry) {
  registry.addAggregate('min', (args) => args.reduce(math.min));
  registry.addAggregate('max', (args) => args.reduce(math.max));
  registry.addAggregate('sum', (args) => args.fold(0.0, (a, b) => a + b));
  registry.addAggregate('mean', mean);
  registry.addAggregate('stddev', stddev);
}

double mean(List<double> args) => args.fold(0.0, (a, b) => a + b) / args.length;

double stddev(List<double> args) {
  final m = mean(args);
  final sum = args.fold(0.0, (a, b) => a + (b - m) * (b - m));
  return math.sqrt(sum / math.max(1, args.length - 1));
}

File: fixtures/scale/rust/medium/src/funcs/stats.rs
use crate::errors::CalcError;

pub fn stat_dispatch(name: &str, args: &[f64]) -> Result<f64, CalcError> {
    match name {
        "min" => stat_min(args),
        "max" => stat_max(args),
        "mean" => stat_mean(args),
        "sum" => Ok(stat_sum(args)),
        other => Err(CalcError::UnknownFunction(other.to_string())),
    }
}

pub fn stat_sum(args: &[f64]) -> f64 {
    args.iter().sum()
}

pub fn stat_mean(args: &[f64]) -> Result<f64, CalcError> {
    if args.is_empty() {
        return Err(CalcError::DomainError("mean needs at least one argument"));
    }
    Ok(stat_sum(args) / args.len() as f64)
}

pub fn stat_min(args: &[f64]) -> Result<f64, CalcError> {
    args.iter()
        .copied()
        .reduce(f64::min)
        .ok_or(CalcError::DomainError("min needs at least one argument"))
}

pub fn stat_max(args: &[f64]) -> Result<f64, CalcError> {
    args.iter()
        .copied()
        .reduce(f64::max)
        .ok_or(CalcError::DomainError("max needs at least one argument"))
}
